text/3649-macros-named-capture-groups.md

```diff
@@ -0,0 +1,797 @@
+captured. It can be used in any location, but its exact behavior is location-
+dependent.
+
+- Arguments: one required argument, `$group` or `$metavariable`
+- Allowed usage: may be used anywhere within the expansion, but some arguments
+  may be disallowed.
+- Output: this returns the number of times a group or metavariable was
+  captured, with some scoping specifics.
+- Changes from current implementation:
+  - Can take a group as an argument
+  - Functionality combined with `len`
+
+Looking at a group or variable that is more deeply nested will return how many
+of that variable were captured in the current repetition. Looking at a variable
+or group that is less deeply nested will return the total times that group was
+captured.
+
+This can be represented as a simple tree walking algorithm to the _capture_
+tree to determine what gets counted. The starting position in the _expansion_
+determines where to start, and then the following rules are applied:
+
+- If `level($name)` >= `level(expression)` (more deeply nested), walk all
+  descendents, including those of neighbors, and count each `$name`
+- If `level($name) < `level(expression)` (less deeply nested):
+  - Walk the entire tree and count each `$name`
+  - Reject code with an error if `$name` is not an ancestor
+
+```rust
+/* looking at descendents */
+
+// Ok: prints `[oa 3, m 3, i 2, x 2; ob 1, y 1]`
+// Demo printing totals. Expansion is at the root (level 0), so each variable
+// is at a higher level; the entire tree is walked and all instances are counted.
+[
+    oa ${count($outer_a)}, m ${count($middle)}, i ${count($inner)}, x ${count($x)};
+    ob ${count($outer_b)}, y ${count($y)},
+]
+
+// Ok: prints `[1 0 2]`
+[$outer_a( ${count($middle} )*]
+
+// Ok: prints `[2 0 0]`
+[$outer_a( ${count($inner} )*]
+
+// Ok: prints `[2 0 0]`
+[$outer_a( ${count($x} )*]
+
+// Ok: prints `o[m[x 2]] o[] o[m[x 0] m[x 0]]`
+// `$middle[0,0]``"sees" two `$x` captures. `$middle[2,0]` and `$middle[2,1]` both
+// don't see any.
+$outer_a( o[$middle( m[x ${count($x)}] )*] )*
+
+/* looking at ancestors */
+
+// Ok: prints [3 3 3]
+// `count` is used at level 2 (one deeper than `outer_a`), so it sees all
+// `outer_a`s each time.
+[$outer_a( ${count($outer_a)} )*]
+
+// Ok: prints [[3] [] [3 3]]
+// Similar to the above
+[$outer_a( [$middle( ${count($outer_a)} )*] )*]
+
+/* errors */
+
+// Error: trying to count a variable that is neither a descendent or anscestor
+[$outer_a( ${count($y} )*]
+```
+
+TODO: we could relax the final rule and allow counting siblings of ancestors
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+- If [`macro_metavar_expr`] stabilizes before this merges, this will add a
+  duplicate way of using those expressions. If this RFC is accepted,
+  stabilizing only a subset of metavariable expressions that does not conflict
+  should be considered.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
```

Comment:
```suggestion

- In macro metavariable syntax, using named capture groups, we could treat `count` and `index` as *fields* rather than *functions*. For instance, we could write `${$group.index}` rather than `${index($group)}`. This would be consistent with the [macro fragment fields](https://github.com/rust-lang/rfcs/pull/3714) proposal.
- Since we no longer need both `count` and `len`, we could choose to use either name for the remaining function we still provide. We should consider whether `count` or `len` best describes this functionality.
```